Fix integration test failure

jshaughn opened the pull request:
Fixes https://github.com/kiali/kiali/issues/6097
Fixes https://github.com/kiali/kiali/issues/6051

Handle the fact that we currently need to equate "" with "Kubernetes" when querying for Namespaces for the home cluster.  This should change as we better handle the home cluster name.

---

@hhovsepy,  I'm not exactly sure why it affects the OCP intergration tests and not other situations, but the root cause is what we had talked about.  The client call 
is using the configured home cluster name, which defaults to "Kubernetes", but the cache is using "".  It probably has to do with our current home cluster handling; I can see 3 different ways of determining the home cluster name: `kubernetes.HomeClusterName`, `business.DefaultClusterID`, and `kialiConfig.Get().KubernetesConfig.ClusterName`.  

For now, I think this resolves the problems, but I think the config is probably the thing that should always be used for the hoce cluster (and it should default to "Kubernetes").

hhovsepy commented on business/namespaces.go:
```diff
@@ -404,7 +404,11 @@ func (in *NamespaceService) GetNamespacesForCluster(ctx context.Context, cluster
+		// or a HomeClusterName constant.  We should always be using the configured cluster name (which should be
+		// set the proper "Kubernetes" default as needed).  The RHS is here because there are times when the client
+		// is using one value and the server is using the other.
+		if ns.Cluster == cluster || (cluster == DefaultClusterID && ns.Cluster == "") {
```
DefaultClusterID is hardcoded to "Kubernetes", but in fact "install-istio-via-istioctl.sh" can take any cluster name, and in QE env it is set "cluster-default", so the tests are still failing.
Suggestion to change 'DefaultClusterID' here to 'config.Get().KubernetesConfig.ClusterName'

hhovsepy replied:
making changes as experimental in https://github.com/kiali/kiali/pull/6101 and checking on OCP cluster, now more failures, investigating

hhovsepy replied:
@jshaughn if you make the changes then we can merge this PR, it blocks other PRs.

jshaughn replied:
Yes, the tests were passing on my system because DefaultClusterID was correct, because I was using the default.  But the main issue here is bigger, we are going to get into trouble with these three different things going on.

jshaughn replied:
I'm going to look at this a bit more.  I still think the solution is to get the proper home cluster name into the cache.  It should not be "".

jshaughn commented:
Hi @hhovsepy , So after diving into this some more I found the root cause of the failures on OCP.  It had to do with openshift projects not being assigned cluster information in the same way that we already had for vanilla namespaces.  I've updated this PR to remove my original band-aid fix and it now has the proper fix.  I went further, though, because I wanted to help make the namespace management a little more clean.  The `kubernetes.HomeClusterName` variable was deprecated and I think that was a good idea.  So I removed it in favor of relying on the config.  I found a variety of other small issues that I cleaned up along the way, some in the code and some in the tests.

I am pretty sure we want to merge this stuff in and layer any other changes on top of it.  Don't be scared by the number of files, most of them are just touched by the removal of HomeClusterName.  I'll be online tomorrow morning for a while, so we can discuss it further if necessary.

**Update** For some reason the CI is failing, but I'm not sure why, I think maybe there is actually a CI env issue because I'm seeing timeouts and things like that.  It's not failing for me locally.  Or maybe it now fails on non-OCP :-(  Anyway, please take a look...

hhovsepy commented:
Thank you Jay, I will take a look,
we had a separate task for optimizing kubernetes.HomeClusterName https://github.com/kiali/kiali/issues/6051, (I was postponing it's solution :) ) so this PR resolves that.

hhovsepy commented:
Graph page is failing to load, when `/graph` request does not contain `cluster` param, it returns '"Cluster [cluster-default] is not found or is not accessible for Kiali"' error response.
But `/graph?cluster=Kubernetes` requests works fine.

hhovsepy commented on business/namespaces.go:
```diff
@@ -556,15 +557,23 @@ func (in *NamespaceService) GetNamespaceByCluster(ctx context.Context, namespace
 			if err2 != nil {
 				return nil, err2
 			}
+		} else {
```
here is one of the key changes, the issue described here https://github.com/kiali/kiali/pull/6106  is resolved now

jshaughn commented on business/tls.go:
```diff
@@ -44,13 +44,15 @@ func (in *TLSService) MeshWidemTLSStatus(ctx context.Context, namespaces []strin
 
-	istioConfigList, err := in.businessLayer.IstioConfig.GetIstioConfigListPerCluster(ctx, criteria, cluster)
+	// @TODO hardcoded HomeClusterName
+	istioConfigList, err := in.businessLayer.IstioConfig.GetIstioConfigListPerCluster(ctx, criteria, conf.KubernetesConfig.ClusterName)
```
I changed this as you suggested.  But I am wondering what is the difference between what is supplied in the `criteria.cluster` and `cluster` arguments?  Also, why wouldn't this `cluster` argument be in the criteria?

hhovsepy replied:
Thank you Jay, it seems separate param remains because the Criteria.cluster was added later, but there is a task to cleanup this code sections: https://github.com/kiali/kiali/issues/6052

hhovsepy reviewed (approved):
Thank you for your work @jshaughn  
On OCP QE single cluster env there is only 1 failures (probably not related).
Tested on local multicluster.
